clippy_lints/src/trivially_copy_pass_by_ref.rs

```diff
@@ -0,0 +1,135 @@
+/// registers.
+///
+/// **Known problems:** This lint is target register size dependent, it is
+/// limited to 32-bit to try and reduce portability problems between 32 and
+/// 64-bit, but if you are compiling for 8 or 16-bit targets then the limit
+/// will be different.
+///
+/// The configuration option `trivial_copy_size_limit` can be set to override
+/// this limit for a project.
+///
+/// **Example:**
+/// ```rust
+/// fn foo(v: &u32) {
+///     assert_eq!(v, 42);
+/// }
+/// // should be
+/// fn foo(v: u32) {
+///     assert_eq!(v, 42);
+/// }
+/// ```
+declare_clippy_lint! {
+    pub TRIVIALLY_COPY_PASS_BY_REF,
+    perf,
+    "functions taking small copyable arguments by reference"
+}
+
+pub struct TriviallyCopyPassByRef {
+    limit: Option<u64>,
+}
+
+impl TriviallyCopyPassByRef {
+    pub fn new(limit: Option<u64>) -> Self {
+        Self { limit }
+    }
+}
+
+impl LintPass for TriviallyCopyPassByRef {
+    fn get_lints(&self) -> LintArray {
+        lint_array![TRIVIALLY_COPY_PASS_BY_REF]
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TriviallyCopyPassByRef {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'a, 'tcx>,
+        kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl,
+        _body: &'tcx Body,
+        span: Span,
+        node_id: NodeId,
+    ) {
+        if in_macro(span) {
+            return;
+        }
+
+        match kind {
+            FnKind::ItemFn(.., abi, _, attrs) => {
+                if abi != Abi::Rust {
+                    return;
+                }
+                for a in attrs {
+                    if a.meta_item_list().is_some() && a.name() == "proc_macro_derive" {
+                        return;
+                    }
+                }
+            },
+            FnKind::Method(..) => (),
+            _ => return,
+        }
+
+        // Exclude non-inherent impls
+        if let Some(NodeItem(item)) = cx.tcx.hir.find(cx.tcx.hir.get_parent_node(node_id)) {
+            if matches!(item.node, ItemImpl(_, _, _, _, Some(_), _, _) |
+                ItemTrait(..))
+            {
+                return;
+            }
+        }
+
+        let fn_def_id = cx.tcx.hir.local_def_id(node_id);
+
+        let fn_sig = cx.tcx.fn_sig(fn_def_id);
+        let fn_sig = cx.tcx.erase_late_bound_regions(&fn_sig);
+
+        for (input, &ty) in decl.inputs.iter().zip(fn_sig.inputs()) {
+            // All spans generated from a proc-macro invocation are the same...
+            if span == input.span {
+                return;
+            }
+
+            let limit = self.limit.unwrap_or_else(|| {
+                // if the config doesn't specify an explicit limit, then limit
+                // it to 2 * the register size, capped at 32-bit to retain
+                // portability between 32 and 64-bit.
+                let bit_width = cx.sess().target.usize_ty.bit_width().unwrap_or(32);
+                cmp::min(bit_width as u64 / 8, 4) * 2
```

Comment:
I don't think the doc comment above is right. aren't you capping it at 64 bits? `min(x, 4) * 2` is ≤ 8 bytes = 64 bit

Reply:
I'll try and reword the comment a bit, it caps the _register size_ at 32-bits, then doubles that to get the maximum size of a type that should be passed by value.